src/common/webrequest_curl.cpp

```diff
@@ -362,217 +510,1194 @@ void wxWebAuthChallengeCURL::SetCredentials(const wxWebCredentials& cred)
-        if ( curl_global_init(CURL_GLOBAL_ALL) )
-            wxLogError(_("libcurl could not be initialized"));
-    }
+        INVALID_ACTION = 0x00,
+        POLL_FOR_READ = 0x01,
+        POLL_FOR_WRITE = 0x02
+    };
 
-    ms_activeSessions++;
+    enum Result
+    {
+        INVALID_RESULT = 0x00,
+        READY_FOR_READ = 0x01,
+        READY_FOR_WRITE = 0x02,
+        HAS_ERROR = 0x04
+    };
+
+    SocketPoller(wxEvtHandler*);
+    ~SocketPoller();
+    bool StartPolling(wxSOCKET_T, int);
+    void StopPolling(wxSOCKET_T);
+    void ResumePolling(wxSOCKET_T);
+
+private:
+    SocketPollerImpl* m_impl;
+};
+
+wxDEFINE_EVENT(wxEVT_SOCKET_POLLER_RESULT, wxThreadEvent);
+
+class SocketPollerImpl
+{
+public:
+    virtual ~SocketPollerImpl(){};
+    virtual bool StartPolling(wxSOCKET_T, int) = 0;
+    virtual void StopPolling(wxSOCKET_T) = 0;
+    virtual void ResumePolling(wxSOCKET_T) = 0;
+
+    static SocketPollerImpl* Create(wxEvtHandler*);
+};
+
+SocketPoller::SocketPoller(wxEvtHandler* hndlr)
+{
+    m_impl = SocketPollerImpl::Create(hndlr);
 }
 
-wxWebSessionCURL::~wxWebSessionCURL()
+SocketPoller::~SocketPoller()
 {
-    {
-        // Notify the work thread
-        wxMutexLocker lock(m_mutex);
-        m_shuttingDown = true;
-        m_condition.Signal();
-    }
+    delete m_impl;
+}
 
-    // Wait for work thread to finish
-    if ( GetThread() && GetThread()->IsRunning() )
-        GetThread()->Wait(wxTHREAD_WAIT_BLOCK);
+bool SocketPoller::StartPolling(wxSOCKET_T sock, int pollAction)
+{
+    return m_impl->StartPolling(sock, pollAction);
+}
+void SocketPoller::StopPolling(wxSOCKET_T sock)
+{
+    m_impl->StopPolling(sock);
+}
 
-    if ( m_handle )
-        curl_multi_cleanup(m_handle);
+void SocketPoller::ResumePolling(wxSOCKET_T sock)
+{
+    m_impl->ResumePolling(sock);
+}
 
-    // Global CURL cleanup if this is the last session
-    --ms_activeSessions;
-    if ( ms_activeSessions == 0 )
-        curl_global_cleanup();
+#ifdef __WINDOWS__
+
+class WinSock1SocketPoller: public SocketPollerImpl
+{
+public:
+    WinSock1SocketPoller(wxEvtHandler*);
+    virtual ~WinSock1SocketPoller();
+    virtual bool StartPolling(wxSOCKET_T, int) wxOVERRIDE;
+    virtual void StopPolling(wxSOCKET_T) wxOVERRIDE;
+    virtual void ResumePolling(wxSOCKET_T) wxOVERRIDE;
+
+private:
+    static LRESULT CALLBACK MsgProc(HWND hwnd, WXUINT uMsg, WXWPARAM wParam,
+                                    WXLPARAM lParam);
+    static const WXUINT SOCKET_MESSAGE;
+
+    WX_DECLARE_HASH_SET(wxSOCKET_T, wxIntegerHash, wxIntegerEqual, SocketSet);
+
+    SocketSet m_polledSockets;
+    WXHWND m_hwnd;
+};
+
+const WXUINT WinSock1SocketPoller::SOCKET_MESSAGE = WM_USER + 1;
+
+WinSock1SocketPoller::WinSock1SocketPoller(wxEvtHandler* hndlr)
+{
+    // Initialize winsock in case it's not already done.
+    WORD wVersionRequested = MAKEWORD(1,1);
+    WSADATA wsaData;
+    WSAStartup(wVersionRequested, &wsaData);
+
+    // Create a dummy message only window.
```

Comment:
We could also use `wxCreateHiddenWindow()` here if we extended it with `hWndParent = 0` parameter.